jowhee3011 iP

jowhee3011 opened the pull request:
# ECHOA README
Task List Status 👍
- [x] OOP
- [x] Different functions for the chat bot
- [x] Debugging

All classes are packaged in Echoa with their relevant import statements
```ruby
package Echoa
import ...
import ...
```

## Summary of Implementations
1. Task Instructions
2. Parsing
3. Task, ToDo, Deadline and Event Classes
4. Task List, Storage and UI Classes
5. Exceptions

### Task Instructions
- todo
- deadline
- event
- list
- delete
- find
- mark
- unmark

Command Line Input format:
```
list
delete [index]
mark [index]
unmark [index]
find [keyword]
```

### Parsing
A Parser Class was implemented to encapsulate all fields and methods related to parsing command line inputs provided by the user.
The Parser can:
- Parse for **instructions**
- Parse for **index**
- Parse for find **keywords**
- Parse for todo, deadline and event **details**

### Task, ToDo, Deadline and Event Classes
ToDo, Deadline and Event Classes are subclasses of the Task class
Task contains the field description.

Deadline contains the following additional fields.
- Due date

Event contains the following additional fields.
- Start Date and Time
- End Date and Time

To input a ToDo task, key ``todo [task description]``
To input a Deadline task, key ``deadline [task description] /[YYYY-MM-DD HH:MM]``
To input an Event task, key ``event [task description] / [YYYY-MM-DD HH:MM] / [YYYY-MM-DD HH:MM]``
### TaskList, Storage and UI Classes
TaskList is a class which encapsulates _tasks in an ArrayList_.
It also contains methods associated to the TaskList.

Storage is a class which deals with the _files used by the program_.

Ui Class deals with the _user interface_.
### Exceptions
Some Exception Classes were implemented to categorize the different errors that could be thrown when there are parsing errors.
- InvalidTaskContentException
- InvalidToDoContentException
- InvalidDeadlineContentException
- InvalidEventContentException
- InvalidIndexInputException
- ListOutOfBoundsException

## References
[Github Basic Writing and Formatting Syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

jowhee3011 committed:
Renamed and edited Echoa.java

jowhee3011 committed:
Allowed Echoa to echo user inputs

jowhee3011 committed:
Edited Echoa to allow adding and listing of tasks

jowhee3011 committed:
Created Task class, modified Echoa to take in mark and unmark and some display changes

jowhee3011 committed:
Added subclasses ToDo, Deadline and Event, modified Echoa to accept different commands

jowhee3011 committed:
Added exception classes and error handling into Echoa

jowhee3011 committed:
Modified structure used for tasklist, added delete instruction

jowhee3011 committed:
Added testing

jowhee3011 committed:
Edit intro line

jowhee3011 committed:
Add file storage

jowhee3011 committed:
Merge branch 'branch-Level-7'

jowhee3011 committed:
Add time and date formatting

jowhee3011 committed:
Merge branch 'branch-Level-8'

jowhee3011 committed:
Add OOP

jowhee3011 committed:
Add package

jowhee3011 committed:
Merge branch 'add-gradle-support'

jowhee3011 committed:
Add JUnit test cases

jowhee3011 committed:
Modify build file with jar names

jowhee3011 committed:
Add JavaDocs

jowhee3011 committed:
Merge branch 'branch-A-JavaDoc'

jowhee3011 committed:
Adjust to coding standards

jowhee3011 committed:
Merge branch 'branch-A-CodingStandard'

jowhee3011 committed:
Add find function

jowhee3011 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	data/echoa.txt
#	src/main/java/Echoa/Echoa.java
#	src/main/java/Echoa/Parser.java
#	src/main/java/Echoa/Ui.java

jowhee3011 committed:
Fix wrong merge errors whiel resolving merge conflict

jowhee3011 committed:
Debug running errors
Add additional exceptions

flyingsalsa commented on src/main/java/Echoa/Deadline.java:
```diff
@@ -0,0 +1,71 @@
+package Echoa;
```
Do note that package names should be in lower case.

flyingsalsa commented on src/main/java/Echoa/Echoa.java:
```diff
@@ -0,0 +1,125 @@
+                        break;
+                    case "":
+                        throw new InvalidInstructionException("Blank");
+                    default:
```
Do note the switch case indentation convention adopted by this course which you can find back in the week 3 videos (3.5) or in the [coding conventions](https://se-education.org/guides/conventions/java/intermediate.html#layout)

DesSnowy replied:
The coding convention for this course require `case` to be on the same indentation as `switch`.

flyingsalsa commented on src/main/java/Echoa/Parser.java:
```diff
@@ -0,0 +1,207 @@
+     * @param task task line.
+     * @return an array containing description.
+     * @throws InvalidToDoContentException when the input format for todo is incorrect.
+     */
```
Good job on properly including all the javaDoc comments!

flyingsalsa commented on src/main/java/Echoa/Storage.java:
```diff
@@ -0,0 +1,135 @@
+        FileWriter fw2 = new FileWriter(file, true);
+        for (int i = 1; i < taskList.getSize(); i++) {
+            fw2.write(taskList.getSpecificTask(i).toText() + System.lineSeparator());
+
```
Did you mean to include this extra line break here?

flyingsalsa commented on src/main/java/Echoa/TaskList.java:
```diff
@@ -0,0 +1,100 @@
+public class TaskList {
+    ArrayList<Task> taskList = null;
+
+    TaskList() {
+        this.taskList = new ArrayList<>();
```
Should there be no modifier for this constructor?

flyingsalsa commented on src/main/java/Echoa/Parser.java:
```diff
@@ -0,0 +1,207 @@
+                tasks.addTask(t);
+            }
+        }
+
+        return tasks;
+    }
```
I've noticed sometimes your return statement has an extra line break to separate it but other times it does not. Have you decided on which one you want to follow?

flyingsalsa reviewed (commented):
Overall coding standards are maintained very well throughout! I could only spot minor errors

DesSnowy commented on src/main/java/Echoa/Event.java:
```diff
@@ -0,0 +1,87 @@
+     * @param date LocalDate to be reformatted.
+     * @return String representation of reformatted date.
+     */
+    public static String getReformattedDate(LocalDate date) {
+        return date.getMonth() + " " +
+               date.getDayOfMonth() + " " + date.getYear();
+    }
```
I noticed the same method on the Deadline class as well. Would it be better if it was placed on the Task class since its a static method?

DesSnowy commented on src/main/java/Echoa/InvalidEventContentException.java:
```diff
@@ -0,0 +1,19 @@
+        super();
+    }
+
+    public String toString() {
+        return
+        "Event has been inputted in the wrong format.\n" +
+        "Please key in the following format:\n" +
+        "event [task name] /[start day, time and/or date] /[start day, time and/or date]";
+    }
```
I noticed most Exception uses the `getMessage` method instead. Perhaps changing this method to follow the rest might be better.

DesSnowy commented on src/main/java/Echoa/InvalidInstructionException.java:
```diff
@@ -0,0 +1,27 @@
+        String message = this.getInstruction() + " is not a valid command!\n" +
+                            "Here are the valid instructions: \n";
+        for (String i : Ui.INSTRUCTION_LIST) {
+            message = message + "- " + i + "\n";
```
You can consider using a StringBuilder instead to reduce runtime cost when the instruction list scales up.

DesSnowy commented on src/main/java/Echoa/InvalidTaskContentException.java:
```diff
@@ -0,0 +1,23 @@
+    @Override
+    public String getMessage() {
+        return
+        "Task has been inputted in the wrong format.\n" +
+        "Please key in the following format for different instructions.\n" +
+        "TODO:\n" +
+        "todo [task name]\n" +
+        "DEADLINE:\n" +
+        "deadline [task name] /[due day, time, and/or date]\n" +
+        "EVENT\n" +
+        "event [task name] /[start day, time and/or date] /[start day, time and/or date]";
```
Based on the coding standard, we generally break before the `+` operator when wrapping lines.

DesSnowy commented on src/main/java/Echoa/Task.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @return String representation of status icon.
+     */
+    public String getStatusIcon() {
+        return (isDone ? "X" : " ");
+    }
```
Could consider changing this to `private` instead since it is only used within the same class.

DesSnowy reviewed (commented):
Good job with the javadocs and naming.

I've pointed out some places that could be refined even more but LGTM otherwise.

Good job!

jowhee3011 committed:
Add GUI

jowhee3011 committed:
Code is not in adherence to code quality standards

Repetitive code in Deadline and Event
Naming conventions not followed in Ui
Naming conventions (small letters) not followed for package naming
Access modifiers not stated in Tasklist
Redundant code in Tasklist
Excessive exception catches in Main

Reduce repeitive code
Renamed to fit naming conventions
Added and removed required code
Structured inheritance of exceptions

Imrprove readability
Make use of polymorphism

jowhee3011 committed:
No assertion in Storage file

File may be null, causing runtime errors

Asserts that file is not null

It is done to capture programming errors as this is not depending on external factors,
such as user input

jowhee3011 committed:
Merge branch 'branch-A-Assertions'

jowhee3011 committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/echoa/Echoa.java

jowhee3011 committed:
Printing errors in Ui
Echoa merge errors

Not user friendly
Cause errors to program

getDeleteMessage in Ui modified
Additional code and syntax from merging errors removed

jowhee3011 committed:
No update function

Increase user ease of chatbot usage

Implement parsing and individual task methods for updating
Implemented methods for updateMessage in Ui
Implemented additional exceptions for updating errors

Make use of polymorphism
Ensure high level of abstraction

jowhee3011 committed:
Update user and echoa image

jowhee3011 committed:
Clean up code

Remove long methods, input javadocs, modify code

Apply SLAP, increase readability, enhance user experience

jowhee3011 committed:
Add Enums

Increase ease of parsing instructions

jowhee3011 committed:
Remove long method in Echoa

Long method is Echoa

Require change as it does not adhere to code quality

Abstract further to allow use of SLAP

Done this way to make use of polymorphism

jowhee3011 committed:
Implement unit testing

jowhee3011 committed:
Debug Parser and edit testing

jowhee3011 committed:
Debug update parses

jowhee3011 committed:
Modify error handling message

jowhee3011 committed:
Give credit through comments

jowhee3011 committed:
Set up product website

jowhee3011 committed:
Add GUI png and user guide

jowhee3011 committed:
Debug removal of index from input string

jowhee3011 committed:
Add JavaDocs

jowhee3011 committed:
Add AI declaration

jowhee3011 committed:
Debug parser
